BinaryTree.java

```diff
@@ -15,99 +24,133 @@ public BinaryTree(BinaryTreeNode<T> root) {
-    public BinaryTreeNode<T> findNode(T key) {
-        if (this.isExist(key)) {
-            return findNode(root, key);
+            return root;
         }
-        System.out.println("The node doesn't exist");
-        return null;
-    }
-
-    // Recursively searches the node with the key and return it
-    public BinaryTreeNode<T> findNode (BinaryTreeNode<T> node, T key){
+        BinaryTreeNode<T> node = exists(root.getLeftChild(), key);
 
         if (node == null) {
-            return null;
+            node = exists(root.getRightChild(), key);
         }
-
-        if (node.getData().equals(key)) {
-            return node;
-        }
-
-        BinaryTreeNode<T> foundNode = findNode(node.getLeftChild(), key);
-
-        if (foundNode == null) {
-            foundNode = findNode(node.getRightChild(), key);
-        }
-        return foundNode;
+        return node;
     }
 
-    // The function create a left (left or right) to the node with the value parent
-    public void insert(T value, T parent, boolean toLeft) {
 
+    // The function create a left (left or right) to the node with the value parent
+    public void insert(T value, T parent) {
         // Check if this value already exists in the tree
-        if (this.isExist(value)) {
-            System.out.println("This value already exists");
+        if (this.exists(value) != null) {
+            System.err.println("This value already exists");
             return;
         }
 
         // Check if the parent exists in the tree
-        if (!this.isExist(parent)) {
-            System.out.println("This parent isn't exist in the tree");
+        if (this.exists(parent) == null) {
+            System.err.println("This parent isn't exist in the tree");
             return;
         }
-
-        insert(value, this.findNode(parent), toLeft);
+        insert(value, this.exists(parent));
     }
 
     // The function create a left (left or right) to the parent node
-    public void insert(T value, BinaryTreeNode<T> parent, boolean toLeft) {
-
-        if (toLeft) {
+    public void insert(T value, BinaryTreeNode<T> parent) {
+        if (m_insertChild == InsertChild.Left) {
             if (parent.getLeftChild() == null) {
                 parent.setLeftChild(value);
             } else {
-                System.out.println("There is already left child");
+                System.err.println("There is already left child");
                 return;
             }
         } else {
             if (parent.getRightChild() == null) {
                 parent.setRightChild(value);
             } else {
-                System.out.println("There is already right child");
+                System.err.println("There is already right child");
                 return;
             }
         }
     }
-}
+
+    // The function receives a Binary Tree and a key
+    // Prints all the ancestors of the key from the root
+    public List<T> printAncestors(BinaryTree<T> tree, T key) {

```

Comment:
This method is doing two things - prints and returns list. This is considered bad practice, as user of the API cannot get one effect without the other. What if I don't want to print the result but only receive list of ancestors?

It would be much better to have findAncestors (or getAncestors or something similar) that would just return list without printing, and a helper method that would take list as parameter and print it out.

Second thing: is the tree parameter really needed? You already have a tree object - `this`.